### ISSUE TITLE
Rule Propsal: No generic type variable shadowing

### ISSUE THREAD

**Issue by nmain** [typescript-eslint/typescript-eslint#521]
Example code:

```ts
class Foo<T> {
    bar<T>() {
        // here, T refers to the T for bar, not the T for Foo
    }
}
```

A rule could ban instances where a generic type variable is introduced with the same name as one in the enclosing scope.  Related to https://github.com/typescript-eslint/typescript-eslint/issues/54

**Comment by bradzacher** [comment]
ideally this would be patched into the base rule https://eslint.org/docs/rules/no-shadow instead of creating a brand new rule for this case.

**Comment by mysticatea** [comment]
TypeScript has two namespace; variables and types. The same name can exist each other, and it's not shadowing.

The `no-shadow` core rule handles variables. I think that another rule is needed for types.

**Comment by bradzacher** [comment]
Technically for typescript there's 2 and a half scopes - variables, types, and variable types (enums and classes I think are the only two that fit into this).

Correct me if I'm wrong - but `no-shadow` relies upon scope analysis?
@armano2 was working on [rewriting the scope analyser](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/parser/src/analyze-scope.ts) so it correctly tracked typescript's multiple scopes.

If that works correctly, then `no-shadow` should work as expected without any modifications (as would other rules that rely upon scope analysis like `no-unused-vars`, etc).

**Comment by mysticatea** [comment]
No. classes and enums just belong to both at the same time. There is not the third kind of namespaces. After the correct scope analysis arrived, `no-shadow` rule would verify only variables.
